### PR TITLE
Story1

### PR DESCRIPTION
**Pull request by Alina1319**

### CODE REVIEW

File: functions.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    $db->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
+    return $db;
+}
+
+function getPokemons($db): array
+{
+    $query = $db->prepare('SELECT `name`, `pokemon_type`,`height`,`weight`,`ability` FROM `pokedex` JOIN `types` 
+                           ON `pokedex`.`type` = `types`.`id`;');
+    $result = $query->execute();
+    $pokemon = $query->fetchALL();
```
Comment (mporam) [inline functions.php]: should have lowercase `ll` in `fetchAll`

File: functions.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    return $pokemon;
+}
+
+function displayPokemon(array $pokemons): string
+{
+   $output = '<section class="pokedex">';
+   foreach ($pokemons as $pokemon){
+        $output .= '<div class="description">'. '<div class="pokemon-name">'. '<p class="pokemon-heading">'.
```
Comment (mporam) [inline functions.php]: you dont need to concatenate every html element, only the PHP variables:
```
$output .= '<div class="description"><div class="pokemon-name"><p class="pokemon-heading">' . $pokemon['name'] . '</p></div>'.
```

File: pokemons_2021-06-01.sql [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+# Dump of table pokedex
+# ------------------------------------------------------------
+
+DROP TABLE IF EXISTS `pokedex`;
+
+CREATE TABLE `pokedex` (
+  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
+  `name` mediumtext,
+  `type` mediumtext,
```
Comment (mporam) [inline pokemons_2021-06-01.sql]: this could be an int as it only contains a number now

File: pokemons_2021-06-01.sql [hunk 4]
```diff
@@ -0,0 +1,59 @@
+# ************************************************************
```
Comment (mporam) [inline pokemons_2021-06-01.sql]: this looks like an old export, where is the types table?

File: test/functions.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+                    '</div></div><div class="col"><div class="type-class"><p class="weight">Weight: 2</p></div>' .
+                    '<div class="type-class">'.
+                    '<p class="ability">Ability: overgrow</p></div></div></div></section>';
+        $result = displayPokemon($input);
+        $this->assertEquals($expected, $result);
+    }
+    public function testDisplayPokemonFailure()
+    {
+        $input = [];
+        $result = DisplayPokemon($input);
```
Comment (mporam) [inline test/functions.php]: the function has a lowercase `d`

File: test/functions.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        $expected = '<section class="pokedex"></section>';
+        $this->assertEquals($expected, $result);
+    }
+    public function testDisplayPokemonMalformed()
+    {
+        $input = "";
+        $this->expectException(TypeError::class);
+        DisplayPokemon($input);
```
Comment (mporam) [inline test/functions.php]: again, lowercase `d`

File: test/functions.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+                    '<div class="type-class">'.
+                    '<p class="ability">Ability: overgrow</p></div></div></div></section>';
+        $result = displayPokemon($input);
+        $this->assertEquals($expected, $result);
+    }
+    public function testDisplayPokemonFailure()
+    {
+        $input = [];
+        $result = DisplayPokemon($input);
+        $expected = '<section class="pokedex"></section>';
```
Comment (mporam) [inline test/functions.php]: is this what you want your code to do? would it not be better if the function returned a message saying "no pokemon present" or something?

**Review by mporam — CHANGES_REQUESTED** [review]

File: functions.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+   $output = '<section class="pokedex">';
+   foreach ($pokemons as $pokemon){
+        $output .= '<div class="description"><div class="pokemon-name"><p class="pokemon-heading">'.
+                    $pokemon['name']. '</p></div><div class="col"><div class="type-class"><p class="type">Type: '.
+                    $pokemon['pokemon_type'] . '</p></div><div class="type-class"><p class="height">Height: '.
+                    $pokemon['height']. '</p></div></div><div class="col"><div class="type-class"><p class="weight">' .
+                    'Weight: '.$pokemon['weight'].'</p></div><div class="type-class"><p class="ability">'.'Ability: '.
+                    $pokemon['ability']. '</p></div></div></div>';
+   }
+        $output .= '</section>';
```
Comment (nealioa) [inline functions.php]: This is indented too far. Should look like:
```
    foreach ... {
        ...
    }
    $output .= '</section>';
    return $output;
```

File: index.php [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+<head>
+    <meta charset="UTF-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Pokedex</title>
+    <link rel="stylesheet" href="./css/main.css">
+    <link rel="preconnect" href="https://fonts.gstatic.com">
+    <link href="https://fonts.googleapis.com/css2?family=Finger+Paint&family=Roboto:wght@100;300;400;500&display=swap" rel="stylesheet">
+    <link rel="icon" href="https://pngimg.com/uploads/pokeball/pokeball_PNG21.png">
+    <link rel="stylesheet" href="resources/css/normalize.css">
```
Comment (nealioa) [inline index.php]: Put normalize.css before your main.css. This way it won't override any styles you have created.

File: test/functions.php [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        $result = displayPokemon($input);
+        $expected = '<section class="pokedex"></section>';
+        $this->assertEquals($expected, $result);
+    }
+    public function testDisplayPokemonMalformed()
+    {
+        $input = "";
+        $this->expectException(TypeError::class);
+        displayPokemon($input);
+
```
Comment (nealioa) [inline test/functions.php]: Unnecessary blank line.

**Review by nealioa — CHANGES_REQUESTED** [review]

File: index.php [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    <meta charset="UTF-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Pokedex</title>
+    <link rel="stylesheet" href="resources/css/normalize.css">
+    <link rel="stylesheet" href="./css/main.css">
+    <link rel="preconnect" href="https://fonts.gstatic.com">
+    <link href="https://fonts.googleapis.com/css2?family=Finger+Paint&family=Roboto:wght@100;300;400;500&display=swap" rel="stylesheet">
+    <link rel="icon" href="https://pngimg.com/uploads/pokeball/pokeball_PNG21.png">
+
```
Comment (nealioa) [inline index.php]: Unnecessary blank line.

File: index.php [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+        </div>
+    </section>
+<?php
+echo displayPokemon($pokemons);
+?>
+</main>
+</body>
+</html>
+
+
```
Comment (nealioa) [inline index.php]: Unnecessary blank line.

**Review by nealioa — CHANGES_REQUESTED** [review]
A couple of blank lines you could get rid of.

And did you see Mike's comment in testDisplayPokemonFailure() about what displayPokemon() could return if it is given an empty array? It's worth thinking about. (We can chat tomorrow morning if you want.)

File: test/functions.php [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+use PHPUnit\Framework\TestCase;
+
+class functions extends TestCase
+{
+    public function testDisplayPokemonSucces()
+    {
+        $input = [
+            [
+                'name'=>'bulbasaur',
+                'type'=>'grass',
```
Comment (nealioa) [inline test/functions.php]: In `displayPokemon()` the key is `pokemon_type`.

File: test/functions.php [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+                'height'=>'2',
+                'weight'=>'2',
+                'ability'=>'overgrow'
+            ]
+        ];
+        $expected = '<section class="pokedex"><div class="description"><div class="pokemon-name">' .
+                    '<p class="pokemon-heading">bulbasaur</p></div><div class="col"><div class="type-class">' .
+                    '<p class="type">Type: </p></div><div class="type-class"><p class="height">Height: 2</p>' .
```
Comment (nealioa) [inline test/functions.php]: I thought the word grass would be included here.

**Review by nealioa — CHANGES_REQUESTED** [review]

**Review by nealioa — APPROVED** [review]

**Pull request merged**